### ISSUE TITLE
Packages with 4-digit version numbers with trailing 0 can't be added/restored

### ISSUE THREAD

**Issue by inosik** [NuGet/Home#13072]
### NuGet Product Used

dotnet.exe, Visual Studio Package Management UI

### Product Version

.NET SDK 8.0.100

### Impact

It's more difficult to complete my work

### Repro Steps & Context

For historical reasons, some internal packages at our company use a 4-digit version number. We use GitLab Package Registry as our internal feed. It doesn't normalize version numbers with the fourth digit being 0:

![StructuredLogViewer_2023-12-06_10-44-21](https://github.com/NuGet/Home/assets/11393003/cf585c18-ba84-4d78-8f6e-4f1dd3837617)

~~The project which produces this package is a .NET Framework project that uses the modern SDK format. The project file has the `VersionPrefix` property set to `4.4.1.0`. The package was created using `msbuild -t:Pack` with .NET SDK 7.x.~~ The package was created using [Paket](https://fsprojects.github.io/Paket/).

a) Such packages can't be added via `dotnet add package ...` to the project (log: [dotnet-add.log](https://github.com/NuGet/Home/files/13579643/dotnet-add.log))

b) When a `PackageReference` is added manually to the project, it can't be restored:

``` sh
$ dotnet restore
  Determining projects to restore...
  Failed to restore C:\xxx\temp.csproj (in 473 ms).
C:\Program Files\dotnet\sdk\8.0.100\NuGet.targets(156,5): error MSB4181: The "RestoreTask" task returned false but did not log an error. [C:\xxx\temp.csproj]
```

[dotnet-restore.log](https://github.com/NuGet/Home/files/13579585/dotnet-restore.log)

**Comment by zivkan** [comment]
@inosik When I go to an empty directory and do `dotnet new classlib ; dotnet pack -p:VersionPrefix=1.2.3.0`, it creates a package whose version is `1.2.3`, not `1.2.3.0`. Same applies for `msbuild -t:Pack -p:VersionPrefix=1.2.3.0`.

Similarly, in restore, NuGet does the same version normalization. So, I can't figure out what kind of `<PackageReference` would result in NuGet trying to download `4.4.1.0` from the server.

Finally, while I don't yet have any evidence that GitLab's NuGet server implementation is doing anything wrong here, the NuGet team at Microsoft don't run GitLab's package registry service, so we're not in a position to make any changes even if they weren't implementing the NuGet protocol as we have intended.

Can you please provide more detailed reproduction steps, so we can better understand how this problem occurs?

**Comment by inosik** [comment]
> When I go to an empty directory and do dotnet new classlib ; dotnet pack -p:VersionPrefix=1.2.3.0, it creates a package whose version is 1.2.3, not 1.2.3.0. Same applies for msbuild -t:Pack -p:VersionPrefix=1.2.3.0.

Sorry, I just realized that I didn't create the package using dotnet/MSBuild but using Paket. I updated the opening post.

> Similarly, in restore, NuGet does the same version normalization. So, I can't figure out what kind of <PackageReference would result in NuGet trying to download 4.4.1.0 from the server.

It's the other way around: All server URLs have "4.4.1.0" in them. The client normalizes the version and tries to download 4.4.1, which results in HTTP 404.

I'll try to set up a project that you can easily download to reproduce this.

**Comment by inosik** [comment]
I'm unable to reproduce this with publicly available package registries.

GitHub Package Registry normalizes 1.0.0.0 to 1.0.0 on push.

GitLab Package Registry has "1.0.0.0" in the URLs for my test package, but it also seems to "unnormalize" URLs with "1.0.0" to "1.0.0.0". Maybe this will be resolved when our internal GitLab instance gets an update.

This is the best I can provide for now:

> All server URLs have "4.4.1.0" in them. The client normalizes the version and tries to download 4.4.1, which results in HTTP 404.

This is also visible in the attached log file in the opening post:

``` sh
$ dotnet add package {package-id} --version 4.4.1.0 --source https://internal-feed/api/v4/projects/{gitlab-project-id}/packages/nuget/index.json
  Determining projects to restore...
  Writing C:\Users\me\AppData\Local\Temp\tmp1utqcl.tmp
info : X.509 certificate chain validation will use the default trust store selected by .NET for code signing.
info : X.509 certificate chain validation will use the default trust store selected by .NET for timestamping.
info : Adding PackageReference for package '{package-id}' into project 'C:\xxx\temp.csproj'.
info : Restoring packages for C:\xxx\temp.csproj...
info :   CACHE https://internal-feed/api/v4/projects/{gitlab-project-id}/packages/nuget/download/{package-id}/index.json
info :   GET https://internal-feed/api/v4/projects/{gitlab-project-id}/packages/nuget/download/{package-id}/4.4.1/{package-id}.4.4.1.nupkg

info :   NotFound https://internal-feed/api/v4/projects/{gitlab-project-id}/packages/nuget/download/{package-id}/4.4.1/{package-id}.4.4.1.nupkg 28ms

info :   GET https://internal-feed/api/v4/projects/{gitlab-project-id}/packages/nuget/download/{package-id}/index.json
info :   OK https://internal-feed/api/v4/projects/{gitlab-project-id}/packages/nuget/download/{package-id}/index.json 23ms
info :   CACHE https://internal-feed/api/v4/projects/{gitlab-project-id}/packages/nuget/download/{package-id}/index.json
info : Package '{package-id}' is compatible with all the specified frameworks in project 'C:\xxx\temp.csproj'.
error: Value cannot be null. (Parameter 'version')
```

(I had to attach a file because I ran into GitHub max. length)

**Comment by inosik** [comment]
Maybe I got another clue:

``` text
info : Restoring packages for C:\xxx\temp.csproj...
info :   CACHE https://internal-feed/api/v4/projects/{gitlab-project-id}/packages/nuget/download/{package-id}/index.json
...
```

I noticed that newly added packages can't be added immediately after they have been pushed. Does NuGet fall back to constructing URLs by itself instead of following links, if it doesn't find the version in the metadata index of the package?

**Comment by zivkan** [comment]
If you check the `nuspec` file in the package created by Paket, does it have a `.0` revision (4th segment)?  I'm trying to understand if this is solely a GitLab Package Registry issue, or a combination of Paket creating packages different to NuGet's official tooling, creating a scenario that GitLab might not have tested.

> GitLab Package Registry has "1.0.0.0" in the URLs for my test package, but it also seems to "unnormalize" URLs with "1.0.0" to "1.0.0.0".

Does this happen only for packages whose version in the `.nuspec` file says `1.0.0.0`, or does it happen even if the package itself specifies `1.0.0`?  If GitLab normalizes to System.Version/Microsoft style 4 segments for all packages, I don't understand how it ever works 😕 

> I noticed that newly added packages can't be added immediately after they have been pushed. Does NuGet fall back to constructing URLs by itself instead of following links, if it doesn't find the version in the metadata index of the package?

No, the way Restore works (well, for PackageReference anyway), is that it first downloads the [list of package versions](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource#enumerate-package-versions), which is this cached `index.json` you see. Once NuGet knows the version it wants to download exists on a server, then it [downloads the package](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource#download-package-content-nupkg). The protocol docs explicitly say:

>> The LOWER_VERSION is the desired package version normalized using NuGet's version [normalization rules](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#normalized-version-numbers).

So, any server not normalizing the version, according to NuGet's version normalization rules, is violating the publicly documented protocol specification.

Anyway, NuGet does cache HTTP responses for 30 minutes, which benefits everyone not trying to use a package published less than 30 minutes ago. If you want to immediately use a package that was just published (assuming also that NuGet has recently downlaoded that version list), then you can run `dotnet nuget locals http-cache --clear`, and then you'll be able to restore the new package version.

**Comment by inosik** [comment]
> If you check the nuspec file in the package created by Paket, does it have a .0 revision (4th segment)? 

Yes, the .nuspec file of the real package has "4.4.1.0" as the version. 

> Does this happen only for packages whose version in the .nuspec file says 1.0.0.0, or does it happen even if the package itself specifies 1.0.0?

So far this only happens with this particular legacy package. All of our other packages follow the Semver spec and have a 3-component version number.

> Once NuGet knows the version it wants to download exists on a server, then it [downloads the package](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource#download-package-content-nupkg). The protocol docs explicitly say:
>
> > The LOWER_VERSION is the desired package version normalized using NuGet's version [normalization rules](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#normalized-version-numbers).
>
> So, any server not normalizing the version, according to NuGet's version normalization rules, is violating the publicly documented protocol specification.

Thank you for clarifying that, that makes sense. GitLab.com doesn't seem to normalize the version number on upload, but seems to understand requests to URLs with normalized version numbers. For example see this URL:

``` text
https://gitlab.com/api/v4/projects/52864958/packages/nuget/metadata/demolib/1.0.0.json
```

Our internal GitLab instance is a bit outdated and maybe this was fixed in the meantime.

**Comment by nkolev92** [comment]
@inosik have you attempted to upgrade your GitLab instance? 

If I'm interpreting the discussion correctly, I don't think there's an action from NuGet at this point.

**Comment by inosik** [comment]
No, not yet. But I'll ask our IT department to do that. Indeed, this looks like it was fixed in GitLab version 16.4.0. Ironically, we're on 16.3.4.

> The LOWER_VERSION is the desired package version **normalized using NuGet's version [normalization rules](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#normalized-version-numbers)**.

This was the important part I wasn't aware of and explains the error I'm getting.
